File: Random/merge_btw_zero.cpp

//Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
 
class Solution {
    public:
        ListNode* mergeNodes(ListNode* head) {
            int merge = 0;
            for(auto *ptr = head, *ptr2 = head->next; ptr2 != nullptr; ptr2 = ptr2->next){
                if(ptr2->val == 0){
                    ptr = ptr->next;
                    ptr->val = merge;
                    
                    merge = 0;
                } else{
                    merge += ptr2->val;
                }
            }
            return head->next;
        }
};

int main(int argc, char const *argv[])
{
    
    ListNode l = ListNode();
    l.next = new ListNode(3);
    l.next->next = new ListNode(1);
    l.next->next->next = new ListNode();
    l.next->next->next->next = new ListNode(4);
    l.next->next->next->next->next = new ListNode(5);
    l.next->next->next->next->next->next = new ListNode(2);
    l.next->next->next->next->next->next->next = new ListNode();


    return 0;
}

File: Random/Merge_btw_zero.py
# Definition for singly-linked list.
from random import betavariate
from tkinter.messagebox import NO
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def mergeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:
        temp2 = head.next
        temp = head
        merge = 0
        

        while temp2.next != None:

            merge += temp2.val
            temp2 = temp2.next

            if temp2.val == 0:
                temp = temp.next
                temp.val = merge
                merge = 0

        temp.next = None
        return head.next


sol = Solution()

h = ListNode()
h.next = ListNode(3)
h.next.next = ListNode(1)
h.next.next.next = ListNode()
h.next.next.next.next = ListNode(4)
h.next.next.next.next.next = ListNode(5)
h.next.next.next.next.next.next = ListNode(2)
h.next.next.next.next.next.next.next = ListNode()

# temp = h
# while temp.next != None:
#     print(temp.val)
#     temp = temp.next


ans = sol.mergeNodes(h)

while ans:
    print(ans.val)
    ans = ans.next